// @mat-datetimepicker/moment/esm2022/adapter/index.mjs
import { NgModule } from '@angular/core';
import { MatMomentDateModule, MomentDateModule, } from '@angular/material-moment-adapter';
import { DatetimeAdapter, MAT_DATETIME_FORMATS, } from '@mat-datetimepicker/core';
import { MomentDatetimeAdapter } from './moment-datetime-adapter';
import { MAT_MOMENT_DATETIME_FORMATS } from './moment-datetime-formats';
import * as i0 from "@angular/core";
export * from './moment-datetime-adapter';
export * from './moment-datetime-formats';
export class MomentDatetimeModule {
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: MomentDatetimeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    /** @nocollapse */ static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.0.4", ngImport: i0, type: MomentDatetimeModule, imports: [MomentDateModule] }); }
    /** @nocollapse */ static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: MomentDatetimeModule, providers: [
            {
                provide: DatetimeAdapter,
                useClass: MomentDatetimeAdapter,
            },
        ], imports: [MomentDateModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: MomentDatetimeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [MomentDateModule],
                    providers: [
                        {
                            provide: DatetimeAdapter,
                            useClass: MomentDatetimeAdapter,
                        },
                    ],
                }]
        }] });
export class MatMomentDatetimeModule {
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: MatMomentDatetimeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    /** @nocollapse */ static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.0.4", ngImport: i0, type: MatMomentDatetimeModule, imports: [MomentDatetimeModule, MatMomentDateModule] }); }
    /** @nocollapse */ static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: MatMomentDatetimeModule, providers: [
            { provide: MAT_DATETIME_FORMATS, useValue: MAT_MOMENT_DATETIME_FORMATS },
        ], imports: [MomentDatetimeModule, MatMomentDateModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: MatMomentDatetimeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [MomentDatetimeModule, MatMomentDateModule],
                    providers: [
                        { provide: MAT_DATETIME_FORMATS, useValue: MAT_MOMENT_DATETIME_FORMATS },
                    ],
                }]
        }] });
