// @mat-datetimepicker/core/esm2022/datetimepicker/datetimepicker.module.mjs
import { A11yModule } from '@angular/cdk/a11y';
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MatIconModule } from '@angular/material/icon';
import { MatDatetimepickerCalendarComponent } from './calendar';
import { MatDatetimepickerCalendarBodyComponent } from './calendar-body';
import { MatDatetimepickerClockComponent } from './clock';
import { MatDatetimepickerComponent, MatDatetimepickerContentComponent, } from './datetimepicker';
import { MatDatetimepickerInputDirective } from './datetimepicker-input';
import { MatDatetimepickerToggleComponent } from './datetimepicker-toggle';
import { MatDatetimepickerMonthViewComponent } from './month-view';
import { MatDatetimepickerYearViewComponent } from './year-view';
import { MatDatetimepickerMultiYearViewComponent } from './multi-year-view';
import { MatButtonModule } from '@angular/material/button';
import { MatDialogModule } from '@angular/material/dialog';
import * as i0 from "@angular/core";
export class MatDatetimepickerModule {
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: MatDatetimepickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    /** @nocollapse */ static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.0.4", ngImport: i0, type: MatDatetimepickerModule, declarations: [MatDatetimepickerCalendarComponent,
            MatDatetimepickerCalendarBodyComponent,
            MatDatetimepickerClockComponent,
            MatDatetimepickerComponent,
            MatDatetimepickerToggleComponent,
            MatDatetimepickerInputDirective,
            MatDatetimepickerContentComponent,
            MatDatetimepickerMonthViewComponent,
            MatDatetimepickerYearViewComponent,
            MatDatetimepickerMultiYearViewComponent], imports: [CommonModule,
            MatButtonModule,
            MatDialogModule,
            MatIconModule,
            OverlayModule,
            A11yModule], exports: [MatDatetimepickerCalendarComponent,
            MatDatetimepickerCalendarBodyComponent,
            MatDatetimepickerClockComponent,
            MatDatetimepickerComponent,
            MatDatetimepickerToggleComponent,
            MatDatetimepickerInputDirective,
            MatDatetimepickerContentComponent,
            MatDatetimepickerMonthViewComponent,
            MatDatetimepickerYearViewComponent,
            MatDatetimepickerMultiYearViewComponent] }); }
    /** @nocollapse */ static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: MatDatetimepickerModule, imports: [CommonModule,
            MatButtonModule,
            MatDialogModule,
            MatIconModule,
            OverlayModule,
            A11yModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: MatDatetimepickerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MatButtonModule,
                        MatDialogModule,
                        MatIconModule,
                        OverlayModule,
                        A11yModule,
                    ],
                    declarations: [
                        MatDatetimepickerCalendarComponent,
                        MatDatetimepickerCalendarBodyComponent,
                        MatDatetimepickerClockComponent,
                        MatDatetimepickerComponent,
                        MatDatetimepickerToggleComponent,
                        MatDatetimepickerInputDirective,
                        MatDatetimepickerContentComponent,
                        MatDatetimepickerMonthViewComponent,
                        MatDatetimepickerYearViewComponent,
                        MatDatetimepickerMultiYearViewComponent,
                    ],
                    exports: [
                        MatDatetimepickerCalendarComponent,
                        MatDatetimepickerCalendarBodyComponent,
                        MatDatetimepickerClockComponent,
                        MatDatetimepickerComponent,
                        MatDatetimepickerToggleComponent,
                        MatDatetimepickerInputDirective,
                        MatDatetimepickerContentComponent,
                        MatDatetimepickerMonthViewComponent,
                        MatDatetimepickerYearViewComponent,
                        MatDatetimepickerMultiYearViewComponent,
                    ],
                }]
        }] });
